web/thesauruses/javascript/ECMAScript 2009/functions.json

```diff
@@ -0,0 +1,113 @@
+{
+  "meta": {
+    "language": "javascript",
+    "language_name": "JavaScript",
+    "structure": "functions",
+    "language_version": "ECMAScript 2009"
+  },
+  "concepts": {
+    "void_function_no_parameters": {
+      "name": "Function that does not return a value and takes no parameters",
+      "code": "function method() {}"
+    },
+    "void_function_with_parameters": {
+      "name": "Function that does not return a value and that takes 1 or more defined parameters",
+      "code": [
+        "function method(param1) {}",
+        "function method(param1, param2) {}"
+      ]
+    },
+    "void_function_with_keyword_parameters": {
+      "name": "Function that does not return a value and that takes 0 or more defined keyword parameters",
+      "code": [
+        "function method({param1}) {}",
+        "function method({param1, param2}) {}"
+      ]
+    },
+    "void_function_variable_parameters": {
+      "name": "Function that does not return a value and function that takes an unknown number of parameters",
+      "code": [
+        "function method() {",
+        "  var params = Array.prototype.slice.call(arguments); // convert arguments to an array",
+        "  // do something with params",
+        "}"
+      ]
+    },
+    "return_value_function_no_parameters": {
+      "name": "Function that returns a value and takes no parameters",
+      "code": "function method() { return true; }"
+    },
+    "return_value_function_with_parameters": {
+      "name": "Function that returns a value and takes 1 or more defined parameters",
+      "code": [
+        "function method(param1) { return true; }",
+        "function method(param1, param2) { return true; }"
+      ]
+    },
+    "return_value_function_with_keyword_parameters": {
+      "name": "Function that returns a value and takes 0 or more defined keyword parameters",
+      "code": "function method({param1, param2}) { return true; }"
```

Comment:
Should this be something like `...({param1: "value", param2: "value"})...` ?

Reply:
> Should this be something like `...({param1: "value", param2: "value"})...` ?

No, but param1 and param2 are variables at this scope, so you can do this :
```js
function method({
    param1, 
    param2 = 'world'
}) {
    console.log({param1, param2}); 
    return true; 
}

method({ param1: 'hello' });
```

Reply:
Keyword parameters are different than what I think this is. So it'd be you can name your parameters and then supply them in any order, and I think you're doing default parameters.

So I'm guessing this is actually "not-implemented"

Reply:
I thought this didn't match but I guess it does. Huh!

Reply:
I tried something today that works too which permits to store attributes with a key including "-" (for example) in variables:
```js
function method({
        'long-name-1': param1,
        'long-name-2': param2
    }) {
        console.log({
            param1,
            param2
        });
    }
```